### PR TITLE
[TT-4994] Middleware: ensure that session with applied policies is returned.

### PR DESCRIPTION
**Pull request by matiasinsaurralde**
During `CheckSessionAndIdentityForValidKey`, we call `ApplyPolicies` on a copy of the session object, in the current implementation this modified session isn't returned, but a previous session is returned instead.

Detail (with comments):
```go
        // 1. The session object is retrieved with SessionDetail:
	session, found = t.Spec.AuthManager.SessionDetail(t.Spec.OrgID, key, false)
	if found {
		key = session.KeyID

                 // 2. The object is cloned, the "session" we're referring to in this block is a different one now...
		session := session.Clone()
		session.SetKeyHash(keyHash)
		t.Logger().Info("Recreating session for key: ", t.Gw.obfuscateKey(key))

		if !t.Spec.GlobalConfig.LocalSessionCache.DisableCacheSessionState {
			go t.Gw.SessionCache.Set(cacheKey, session, cache.DefaultExpiration)
		}

                 // 3. We apply policies to this cloned session object:
		if err := t.ApplyPolicies(&session); err != nil {
			t.Logger().Error(err)
			return session, false
		}

		t.Logger().Debug("Lifetime is: ", session.Lifetime(t.Spec.SessionLifetime, t.Gw.GetConfig().ForceGlobalSessionLifetime, t.Gw.GetConfig().GlobalSessionLifetime))
                 // 4. Session update is scheduled.
		ctxScheduleSessionUpdate(r)
	} else {
		// defaulting
		session.KeyID = key
	}

         // 5. The session that's returned here isn't the "cloned" session but the unmodified object from the first step.
	return session, found
```

## Related Issue
[TT-4994](https://tyktech.atlassian.net/browse/TT-4994?atlOrigin=eyJpIjoiYWM5ZDY0OTA4MTE1NDE5NmI1NzZkZGQzNDJhNGM2YTMiLCJwIjoiaiJ9)

## Motivation and Context
To achieve expected behavior.

## How This Has Been Tested
Manually tested with custom scenario provided by QA.

## Screenshots (if appropriate)
-

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3993/merge
Commit:  
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2104129484)

**Review by tbuchaillot — APPROVED** [review]

File: gateway/middleware.go [hunk 1]
```diff
@@ -703,34 +703,34 @@ func (t BaseMiddleware) CheckSessionAndIdentityForValidKey(originalKey string, r
 
 	// Only search in RPC if it's not in emergency mode
 	t.Logger().Debug("Querying authstore")
+
 	// 2. If not there, get it from the AuthorizationHandler
 	session, found = t.Spec.AuthManager.SessionDetail(t.Spec.OrgID, key, false)
-	if found {
-		key = session.KeyID
-
-		session := session.Clone()
```
Comment (buger) [inline gateway/middleware.go:711]: Feels like a potential issue to me, which can cause some concurency related issues. 
We use clone for the purpose, so multiple threads in gw can operate with same session objects (clones), so we will not have issues like map writes from multiple threads and etc. 

I think the valid code here is: `session = session.Clone()`

**Comment by tbuchaillot** [comment]
@matiasinsaurralde @buger  thinking a little bit more about it, I would maintain the one-liner change (returning the session) to be cohesive with the rest of `CheckSessionAndIdentityForValidKey ` func.  In the other storages search, we found it, clone it, and return it.


https://github.com/TykTechnologies/tyk/blob/5d206211fc5032a7d25005311ecc39f99e2ffd4b/gateway/middleware.go#L663https://github.com/TykTechnologies/tyk/blob/5d206211fc5032a7d25005311ecc39f99e2ffd4b/gateway/middleware.go#L677

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3993/merge
Commit: 0a9a7b2e2f08eab1e2ea6211d260959b7948b146 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2108047478)

File: gateway/middleware.go [hunk 2]
```diff
@@ -726,11 +726,11 @@ func (t BaseMiddleware) CheckSessionAndIdentityForValidKey(originalKey string, r
...
 		ctxScheduleSessionUpdate(r)
-	} else {
-		// defaulting
-		session.KeyID = key
+		return session, found
 	}
 
+	// defaulting
+	session.KeyID = key
 	return session, found
```
[the remark this answers is not in the document]
Comment (tbuchaillot) [reply gateway/middleware.go]: updated @buger

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3993/merge
Commit: 8413bed5bc3ae9bca70174027788ddcab368af29 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2109603810)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 6f1253cfac3ba3ddf33126185d14aac7fba2791f [TT-4994] Middleware: ensure that session with applied policies is returned. (#3993)

During `CheckSessionAndIdentityForValidKey`, we call `ApplyPolicies` on a copy of the session object, in the current implementation this modified session isn't returned, but a previous session is returned instead.

Detail (with comments):
```go
        // 1. The session object is retrieved with SessionDetail:
	session, found = t.Spec.AuthManager.SessionDetail(t.Spec.OrgID, key, false)
	if found {
		key = session.KeyID

                 // 2. The object is cloned, the "session" we're referring to in this block is a different one now...
		session := session.Clone()
		session.SetKeyHash(keyHash)
		t.Logger().Info("Recreating session for key: ", t.Gw.obfuscateKey(key))

		if !t.Spec.GlobalConfig.LocalSessionCache.DisableCacheSessionState {
			go t.Gw.SessionCache.Set(cacheKey, session, cache.DefaultExpiration)
		}

                 // 3. We apply policies to this cloned session object:
		if err := t.ApplyPolicies(&session); err != nil {
			t.Logger().Error(err)
			return session, false
		}

		t.Logger().Debug("Lifetime is: ", session.Lifetime(t.Spec.SessionLifetime, t.Gw.GetConfig().ForceGlobalSessionLifetime, t.Gw.GetConfig().GlobalSessionLifetime))
                 // 4. Session update is scheduled.
		ctxScheduleSessionUpdate(r)
	} else {
		// defaulting
		session.KeyID = key
	}

         // 5. The session that's returned here isn't the "cloned" session but the unmodified object from the first step.
	return session, found
```

## Related Issue
[TT-4994](https://tyktech.atlassian.net/browse/TT-4994?atlOrigin=eyJpIjoiYWM5ZDY0OTA4MTE1NDE5NmI1NzZkZGQzNDJhNGM2YTMiLCJwIjoiaiJ9)

## Motivation and Context
To achieve expected behavior.

## How This Has Been Tested
Manually tested with custom scenario provided by QA.

## Screenshots (if appropriate)
-

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist


- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

Co-authored-by: Tomas Buchaillot <tombuchaillot89@gmail.com>
Triggered by: push (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2109618283)

**Pull request merged**